src/rust/engine/options/src/parse.rs

```diff
@@ -150,6 +154,51 @@ peg::parser! {
 
         pub(crate) rule string_list_edits() -> Vec<ListEdit<String>>
             = implicit_add(<unquoted_string()>) / list_replace(<quoted_string()>) / list_edits(<quoted_string()>)
+
+        // Heterogeneous values embedded in dicts. Note that float_val() must precede int_val() so that
+        // the integer prefix of a float is not interpreted as an int.
+        rule val() -> Val
+            = v:(bool_val() / float_val() / int_val() / string_val() / list_val() / tuple_val() / dict_val()) {
+            v
+        }
+
+        rule bool_val() -> Val = x:bool() { Val::Bool(x) }
+        rule float_val() -> Val = x:float() { Val::Float(x) }
+        rule int_val() -> Val = x:int() { Val::Int(x) }
+        rule string_val() -> Val = x:quoted_string() { Val::String(x) }
+        rule list_val() -> Val = items:list_items(<val()>) { Val::List(items) }
+        rule tuple_val() -> Val = items:tuple_items(<val()>) { Val::List(items) }
+        rule dict_val() -> Val = whitespace()* d:dict() { Val::Dict(d) }
+
+        rule dict() -> HashMap<String, Val>
+            = dict_start()
+            items:dict_item() ** ","
+            whitespace()* ","? whitespace()*
+            dict_end()
+            whitespace()* {
+                items.into_iter().collect()
+            }
+
+        rule dict_start() -> ()
+            = quiet!{ "{" }
+            / expected!("the start of a dict indicated by '{' or '+{'")
+
+        rule dict_end() -> ()
+            = quiet!{ "}" }
+            / expected!("the end of a dict indicated by '}'")
+
+        rule dict_item() -> (String, Val)
+            = whitespace()* key:quoted_string() whitespace()* ":" whitespace()* value:val() whitespace()* {
+                (key, value)
+            }
+
+        pub(crate) rule dict_edit() -> DictEdit
+            = whitespace()* plus:"+"? d:dict() {
```

Comment:
It looks like this might not allow `+ {...}` with whitespace, is that intentional?

(Potentially this applies to lists too?)

Reply:
Intentional (and the same with lists). We don't support that in the Python parser either.